File: backend/src/main/kotlin/org/globe42/web/exception/ExceptionConfig.java
package org.globe42.web.exception;

import java.util.Map;

import org.springframework.boot.web.servlet.error.DefaultErrorAttributes;
import org.springframework.boot.web.servlet.error.ErrorAttributes;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.web.context.request.WebRequest;

/**
 * Spring Configuration class that adds a custom {@link ErrorAttributes} bean to the context, in order to
 * add a <code>functionalError</code> attribute to the JSON body of error responses when the exception
 * is a {@link BadRequestException}
 * @author JB Nizet
 */
@Configuration
public class ExceptionConfig {
    @Bean
    public ErrorAttributes errorAttributes() {
        return new CustomErrorAttributes();
    }

    public static class CustomErrorAttributes extends DefaultErrorAttributes {
        public CustomErrorAttributes() {
        }

        public CustomErrorAttributes(boolean includeException) {
            super(includeException);
        }

        @Override
        public Map<String, Object> getErrorAttributes(WebRequest webRequest, boolean includeStackTrace) {
            Map<String, Object> result = super.getErrorAttributes(webRequest, includeStackTrace);

            Throwable error = getError(webRequest);
            if (error instanceof BadRequestException) {
                BadRequestException exception = (BadRequestException) error;
                if (exception.getError() != null) {
                    result.put("functionalError", exception.getError());
                }
            }
            return result;
        }
    }
}

File: backend/src/main/kotlin/org/globe42/web/exception/ExceptionConfig.kt
package org.globe42.web.exception

import org.springframework.boot.web.servlet.error.DefaultErrorAttributes
import org.springframework.boot.web.servlet.error.ErrorAttributes
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.web.context.request.WebRequest

/**
 * Spring Configuration class that adds a custom [ErrorAttributes] bean to the context, in order to
 * add a `functionalError` attribute to the JSON body of error responses when the exception
 * is a [BadRequestException]
 * @author JB Nizet
 */
@Configuration
class ExceptionConfig {
    @Bean
    fun errorAttributes(): ErrorAttributes {
        return CustomErrorAttributes()
    }

    open class CustomErrorAttributes : DefaultErrorAttributes {
        constructor()

        constructor(includeException: Boolean) : super(includeException)

        override fun getErrorAttributes(webRequest: WebRequest, includeStackTrace: Boolean): Map<String, Any> {
            val result = super.getErrorAttributes(webRequest, includeStackTrace)

            val error = getError(webRequest)
            if (error is BadRequestException) {
                val exception = error
                if (exception.error != null) {
                    result["functionalError"] = exception.error
                }
            }
            return result
        }
    }
}
